network/phonebook.go

```diff
@@ -47,50 +47,78 @@ type Phonebook interface {
 	// UpdateConnectionTime will update the provisional connection time.
 	// Returns true of the addr was in the phonebook
 	UpdateConnectionTime(addr string, provisionalTime time.Time) bool
+
+	// ReplacePeerList merges a set of addresses with that passed in for networkName
+	// new entries in dnsAddresses are being added
+	// existing items that aren't included in dnsAddresses are being removed
+	// matching entries don't change
+	ReplacePeerList(dnsAddresses []string, networkName string)
+
+	// ExtendPeerList adds unique addresses to this set of addresses
+	ExtendPeerList(more []string, networkName string)
 }
 
-type phonebookData struct {
+type phonebookEntryData struct {
 	retryAfter            time.Time
 	recentConnectionTimes []time.Time
+	networkNames          map[string]bool
 }
 
-// phonebookEntries holds the server connection configuration values
-// and the list of request times withing the time window for each
+func makePhonebookEntryData(networkName string) phonebookEntryData {
+	pbData := phonebookEntryData{
+		networkNames:          make(map[string]bool),
+		recentConnectionTimes: make([]time.Time, 0),
+	}
+	pbData.networkNames[networkName] = true
+	return pbData
+}
+
+// phonebookImpl holds the server connection configuration values
+// and the list of request times within the time window for each
 // address.
-type phonebookEntries struct {
+type phonebookImpl struct {
 	connectionsRateLimitingCount  uint
 	connectionsRateLimitingWindow time.Duration
-	data                          map[string]phonebookData
+	data                          map[string]phonebookEntryData
+	lock                          deadlock.RWMutex
 }
 
-// makePhonebookEntries creates phonebookEntries with the passed configuration values
-func makePhonebookEntries(connectionsRateLimitingCount uint,
-	connectionsRateLimitingWindow time.Duration) phonebookEntries {
-	return phonebookEntries{
+// MakePhonebook creates phonebookImpl with the passed configuration values
+func MakePhonebook(connectionsRateLimitingCount uint,
+	connectionsRateLimitingWindow time.Duration) Phonebook {
+	return &phonebookImpl{
 		connectionsRateLimitingCount:  connectionsRateLimitingCount,
 		connectionsRateLimitingWindow: connectionsRateLimitingWindow,
-		data:                          make(map[string]phonebookData, 0),
+		data:                          make(map[string]phonebookEntryData, 0),
+	}
+}
+
+func (e *phonebookImpl) deletePhonebookEntry(entryName, networkName string) {
+	pbEntry := e.data[entryName]
+	delete(pbEntry.networkNames, networkName)
+	if 0 == len(pbEntry.networkNames) {
```

Comment:
1. how pbEntry.networkNames can be non zero after deletion?
2. this is looks like the very first usage of "0 ==" order. I'd prefer to keep it consistent with all the rest code

Reply:
Can be non-zero after deletion because we are only deleting networkName from pbEntry.networkNames, and  there could be other elements besides the given networkName there. 

0 == comes from C/C++. avoiding an l-value to the left of a logical operator helps the compiler catch the = instead of == errors. This is not be an issue in go. I'll reverse it.